docs/References/Settings.md
::: ipfabric.settings.attributes

::: ipfabric.settings.authentication

::: ipfabric.settings.seeds

::: ipfabric.settings.site_separation

::: ipfabric.settings.user_mgmt

::: ipfabric.settings.vendor_api

::: ipfabric.settings.vendor_api_models

CHANGELOG.md
# Changelog

## 6.0.0 (2022-10-XX)

### Breaking Change

* Snapshot model has changed to use POST which contains more metadata
* GET is still used currently for some missing data
* Initializing of the IPFClient will only get loaded snapshots by default to
  reduce memory as some customers can have over 100 unloaded snapshots

### Feature

* Updates for v6.0
* Will only initialize loaded only snapshots unless `unloaded=True`
* Update Dev Dependencies
* Update for AWS API for multiple regions/AssumeRoles
* Add Lock/Unlock snapshots
* Added get_snapshot(snapshot_id) to find a snapshot_id

## 5.0.15 (2022-10-05)

### Feature

* Added ipfabric.settings.SiteSeparation
* Removed ipfabric.tools.UpdateSiteNames as this was removed in 4.4
* Added ipfabric.tools.site_seperation_report.map_devices_to_rules
* See [site_separation_tester.py](examples/tools/site_separation_report.py)

## 5.0.14 (2022-10-04)

### Feature

* Added shared tools for parsing MAC addresses and return IP Fabric Format
* Added examples for parsing MAC and Timestamps.

## 5.0.13 (2022-10-03)

### Feature

* Added tools for Discovery History table

## 5.0.11 (2022-09-20)

### Fix

* Poetry Update

### CI/CD

* Add community-fabric Slack notification

## 5.0.11 (2022-09-20)

### Fix

* Final fix for SSL verification

## 5.0.10 (2022-09-12)

### Feature

* added the ability to add custom loggers
* add python-json-logger to poetry optional dependencies

## 5.0.9 (2022-09-09)

### Fix

* Removed pkg_resources as it was causing some issues on certain OS and Python
  versions
* Updated poetry-core

### CI

* Added more python versions to functional testing.

## 5.0.8 (2022-09-08)

### Fix

* Update python-dontenv

## 5.0.7 (2022-09-07)

### Fix

* Fix for httpx timeout

## 5.0.6 (2022-09-01)

### Fix

* Fix for unverified cert

## 5.0.5 (2022-08-17)

### Fix

* Fix in user/pass auth

## 5.0.4 (2022-08-17)

### Fix

* Fix in Users model
* Added get_user to get information about logged in user during init
* Added count function to tables

## 5.0.3 (2022-08-11)

### Fix

* Deprecate IPF v4.X from package
* Added Technologies to the SDK
* Verification of API Version against IP Fabric Version

## 5.0.2 (2022-08-09)

### Fix

* Remove `importlib-metadata` dependency

## 5.0.1 (2022-07-13)

### Fix

* Do not use count for paging results do to some tables not returning a count.

## 5.0.0 (2022-07-13)

### Feature

* Update package to be compatible with v5.0.0
* Added tools.RestoreIntents which will restore Intents, Groups, & Dashboard
  from a file, another instance, or factory default
* Added all inventory endpoints to inventory (phones, fans, modules,
  powerSupplies, powerSuppliesFans)
* Allow initialization without loaded snapshots
* Added IPFClient.loaded_snapshots and IPFClient.unloaded_snapshots
* Added a Load and Unload snapshot function to Snapshot
  model `ipf.snapshots['ea0da479-6222-4cf6-a0ae-78a610669477'].load(ipf)`
* Added ability to get attributes for a
  snapshot `attributes = ipf.snapshots['$last'].attributes(ipf)`

## 4.4.3 (2022-06-22)

### Feature

* Add get_count function

### CI

* Remove semantic-release and updated GH Actions

## v0.11.0 (2022-05-25)

### Feature

* Add username/password
  auth ([`4ae04aa`](https://github.com/community-fabric/python-ipfabric/commit/4ae04aab72f1caa1de897ded32d44c9dd01fd187))
* Add username/password
  auth ([`420b6b1`](https://github.com/community-fabric/python-ipfabric/commit/420b6b17b9c0f9c982acf192ead1bf9f9a374dfc))

## v0.10.8 (2022-04-22)

### Fix

* Test discord
  webhook ([`07c04dd`](https://github.com/community-fabric/python-ipfabric/commit/07c04dd7ec646bb509fc0c580561f4318fdb8374))

## v0.10.7 (2022-04-21)

### Fix

* Update
  docs ([`47049fe`](https://github.com/community-fabric/python-ipfabric/commit/47049fecc5b8bfae151a3b7aabb999e9da685270))

## v0.10.6 (2022-04-21)

### Fix

* Update
  packages ([`8ff5efc`](https://github.com/community-fabric/python-ipfabric/commit/8ff5efcdc4e03d69e64daecc39f9cac48a91f062))

## v0.10.5 (2022-04-21)

### Fix

* Update
  release ([`0fb2627`](https://github.com/community-fabric/python-ipfabric/commit/0fb262789f345ac916235a3457c696e6ecbff62f))
* Update
  release ([`ad00fbd`](https://github.com/community-fabric/python-ipfabric/commit/ad00fbdffbb78eb52b617fd4f49214a377a43acf))

## v0.10.4 (2022-04-21)

### Fix

* Update
  release ([`79e5d0d`](https://github.com/community-fabric/python-ipfabric/commit/79e5d0dc640565b850ba4f89abc49bd9d11cc598))
* Update
  release ([`f2a3079`](https://github.com/community-fabric/python-ipfabric/commit/f2a30799c9e502dd046e978df49de8ddd136b59b))
* Update
  release ([`ba6b599`](https://github.com/community-fabric/python-ipfabric/commit/ba6b599ef54584cdea75a1df3ad3198bfd25a5ba))

## v0.10.3 (2022-04-21)

### Fix

* Update
  release ([`a6fec0c`](https://github.com/community-fabric/python-ipfabric/commit/a6fec0c2fcd3f5a0cd9c7f8bf63d753a6457312d))

## v0.10.2 (2022-04-21)

### Fix

* Update
  black ([`baafd60`](https://github.com/community-fabric/python-ipfabric/commit/baafd607b8e28eb1e0791bc779581064b44323e7))
* Test adding
  vulnerabilities ([`ff0fda1`](https://github.com/community-fabric/python-ipfabric/commit/ff0fda1ae7b27422c147a06047989d41ddee3bdd))
* Config
  again ([`6b3fac9`](https://github.com/community-fabric/python-ipfabric/commit/6b3fac9360831f84224a84083c4e89fe22e4c848))
* Poetry
  update ([`24ca2e9`](https://github.com/community-fabric/python-ipfabric/commit/24ca2e98257531465349e672efc019b11d055031))
* Remove graphing and security as these do not work in v4.3 and
  above. ([`a7b3b35`](https://github.com/community-fabric/python-ipfabric/commit/a7b3b353edf8b1f8df222b12778755df441c6a0c))

## v0.10.1 (2022-04-21)

### Fix

* Updates to fix configs in IPF
  v4.4.1 ([`da3e0b8`](https://github.com/community-fabric/python-ipfabric/commit/da3e0b8241bfd36b57edb6bf3c5ac5066d4fc8b2))

## v0.10.0 (2022-04-21)

### Feature

* Updates to fix configs in IPF

examples/tools/config.py
from ipfabric import IPFClient
from ipfabric.tools import DeviceConfigs


if __name__ == "__main__":
    ipf = IPFClient()
    # ipf = IPFClient('https://demo3.ipfabric.io/', token='token', verify=False, timeout=15)
    cfg = DeviceConfigs(ipf)

    cfgs = cfg.get_all_configurations()  # Get all Configurations
    #  Returns dictionary like {SN: [Config, Config]} where SN is the IPF Unique Serial Number
    L35AC92_cfgs = cfg.get_all_configurations(device='L35AC92')  # Get all Configurations for device
    #  Returns dictionary like {'a23ffbf': [Config, Config], '91624130': [Config, Config]}
    print(L35AC92_cfgs)
    """
    {
    'a23ffbf': [Config(config_id='61f1f52936add003e73b4848', sn='a23ffbf', hostname='L35AC92', 
        config_hash='daf3d34ceb44870cc7e2647ccb01dd759e2da5b9', status='no change', 
        last_change=datetime.datetime(2022, 1, 27, 1, 28, 9, tzinfo=datetime.timezone.utc), 
        last_check=datetime.datetime(2022, 2, 10, 1, 21, 29, 986000, tzinfo=datetime.timezone.utc), text=None)],
    '91624130': [Config(config_id='5ffdd5f93d1f1b0343af6bd3', sn='91624130', hostname='L35AC92', 
        config_hash='6df7ab1f8dd2eefc757e8719421d81643c2eb6ee', status='no change', 
        last_change=datetime.datetime(2021, 1, 12, 17, 1, 45, tzinfo=datetime.timezone.utc), 
        last_check=datetime.datetime(2021, 4, 6, 23, 3, 15, 77000, tzinfo=datetime.timezone.utc), text=None)]
    }
    """
    print()

    text = cfg.get_text_config(L35AC92_cfgs['a23ffbf'][0], False).text
    print(text)
    """
    version 15.2
    service timestamps debug datetime msec
    service timestamps log datetime msec
    service password-encryption
    service compress-config
    !
    hostname L35AC92
    !
    boot-start-marker
    """
    print()

    L35AC92_sn_cfgs = cfg.get_all_configurations(sn='a23ffbf')
    #  Returns dictionary like {'a23ffbf': [Config, Config]}
    print(L35AC92_sn_cfgs)
    """
    {
    'a23ffbf': [Config(config_id='61f1f52936add003e73b4848', sn='a23ffbf', hostname='L35AC92', 
        config_hash='daf3d34ceb44870cc7e2647ccb01dd759e2da5b9', status='no change', 
        last_change=datetime.datetime(2022, 1, 27, 1, 28, 9, tzinfo=datetime.timezone.utc), 
        last_check=datetime.datetime(2022, 2, 10, 1, 21, 29, 986000, tzinfo=datetime.timezone.utc), text=None)]
    }
    """
    print()

    L35AC92 = cfg.get_configuration('L35AC92', sanitized=False)  # Display passwords
    print(L35AC92.last_change)
    """
    2021-11-24 00:13:26+00:00
    """
    print()

    L34R3 = cfg.get_configuration('10.35.189.92', date="$prev")  # Method can also accept any managed IP of the device
    # Date can be $last, $prev, $first; $last is default
    print(L34R3.last_change)
    """
    2021-10-18 23:14:27+00:00
    """
    print()

    # Date can also be tuple of start, end which will return the latest config in that time range.
    # If integer timestamp must be seconds not milliseconds.
    L1LB1 = cfg.get_configuration('L1LB1', date=("11/22/2021 1:30", 1637629200))
    print(L1LB1.last_change)
    print(L1LB1.text)
    """
    auth partition Partition1 { }
    auth partition longDescription {
        description "f sf df\" \"dffddfdf"
    }
    ...
    """
    print()

    bad = cfg.get_configuration("bad-device-name")  # Will return None not error.
    print(bad)
    """
    None
    Could not find a matching device for 'bad-device-name' using regex \
        '^[Bb][Aa][Dd]-[Dd][Ee][Vv][Ii][Cc][Ee]-[Nn][Aa][Mm][Ee]$'.
    """
    print()

    log = cfg.get_log('L1LB1', snapshot_id='$last')
    print(log)
    """
    admin15@(L1LB1)(cfg-sync Standalone)(Active)(/Common)(tmos)# show version
    Syntax Error: "version" unexpected argument
    admin15@(L1LB1)(cfg-sync Standalone)(Active)(/Common)(tmos)# uname -a
    Syntax Error: unexpected argument "uname"
    ...
    """
